model/riscv_vmem.sail

```diff
@@ -0,0 +1,851 @@
+      (Execute(),	User)	    => (pte_U == 0b1) & (pte_X == 0b1),
+      (Read(_),		Supervisor) => ((pte_U == 0b0) | do_sum)
+				       & ((pte_R == 0b1) | ((pte_X == 0b1) & mxr)),
+      (Write(_),	Supervisor) => ((pte_U == 0b0) | do_sum)
+				       & (pte_W == 0b1),
+      (ReadWrite(_, _), Supervisor) => ((pte_U == 0b0) | do_sum)
+				       & (pte_W == 0b1)
+				       & ((pte_R == 0b1)
+					  | ((pte_X == 0b1) & mxr)),
+      (Execute(),	Supervisor) => (pte_U == 0b0) & (pte_X == 0b1),
+      (_,		Machine)    => internal_error(__FILE__, __LINE__,
+						      "m-mode mem perm check")};
+  if b then PTE_Check_Success(())
+  else	    PTE_Check_Failure((), ())
+}
+
+// Update PTE bits if needed; return new PTE if updated
+// PRIVATE
+function update_PTE_Bits(sv_params : SV_Params,
+			 pte	   : bits(64),
+			 a	   : AccessType(ext_access_type),
+			 ext	   : extPte)
+			-> option(bits(64)) = {
+  let pte_lsbs = Mk_PTE_LSBs (pte [7 .. 0]);
+
+  // Update 'dirty' bit?
+  let update_d : bool = (pte_lsbs.D() == 0b0)
+			& (match a {
+			     Execute()	     => false,
+			     Read()	     => false,
+			     Write(_)	     => true,
+			     ReadWrite(_, _) => true
+			   });
+  // Update 'accessed'-bit?
+  let update_a : bool = pte_lsbs.A() == 0b0;
+
+  if update_d | update_a then {
+    let pte_lsbs1 = update_A (pte_lsbs, 0b1);
+    let pte_lsbs2 = if update_d then update_D (pte_lsbs, 0b1) else pte_lsbs1;
+
+    let pte2 = pte | zero_extend (pte_lsbs.bits());
+    Some(pte2 | (ext << sv_params.pte_msbs_lsb_index))
+  }
+  else
+    None()
+}
+
+// ****************************************************************
+// Results of Page Table Walk (PTW)
+
+// 'ext_ptw' supports (non-standard) extensions to the default addr-translation and PTW.
+
+// Failure modes for address-translation/page-table-walks
+// PRIVATE
+union PTW_Error = {
+  PTW_Invalid_Addr  : unit,	     // invalid source address
+  PTW_Access	    : unit,	     // physical memory access error for a PTE
+  PTW_Invalid_PTE   : unit,
+  PTW_No_Permission : unit,
+  PTW_Misaligned    : unit,	     // misaligned superpage
+  PTW_PTE_Update    : unit,	     // PTE update needed but not enabled
+  PTW_Ext_Error	    : ext_ptw_error  // parameterized for errors from extensions
+}
+
+// PRIVATE: only 'to_str' overload is public
+function ptw_error_to_str(e : PTW_Error) -> string = {
+  match e {
+    PTW_Invalid_Addr()	 => "invalid-source-addr",
+    PTW_Access()	 => "mem-access-error",
+    PTW_Invalid_PTE()	 => "invalid-pte",
+    PTW_No_Permission()	 => "no-permission",
+    PTW_Misaligned()	 => "misaligned-superpage",
+    PTW_PTE_Update()	 => "pte-update-needed",
+    PTW_Ext_Error(e)	 => "extension-error"
+  }
+}
+
+// PUBLIC
+overload to_str = {ptw_error_to_str}
+
+// hook for (non-standard) extensions to customize errors reported by page-table
+// walks during address translation; it typically works in conjunction
+// with any customization to check_PTE_permission().
+
+// PRIVATE
```

Comment:
Btw I wonder if we should adopt the "leading underscore = private" convention, given that Sail doesn't support public/private natively yet.

Maybe after modules are added.

(No changes needed here; just a thought).

Reply:
Let’s wait for modules, which I believe will be available soon.